### ISSUE TITLE
ChaosExperiment securityContext not updating

### ISSUE THREAD

**Issue by golthitarun** [litmuschaos/litmus#2863]
**What happened**:
Tried creating a ChaosExperiment custom resource with securityContext as describe in docs here -? https://docs.litmuschaos.io/docs/chaosexperiment/#configuration-specification

```spec:
  definition:
    securityContext:
      podSecurityContext:
        runAsUser: 1000
```

Also tried with containerSecurityContext. But I don't see anything populated in the resulting custom resource created, it always shows up as an empty object. 

`securityContext: {}`

Not sure if I am doing anything wrong.

**What you expected to happen**:

I am expecting the resulting custom resource to have securityContext

**How to reproduce it (as minimally and precisely as possible)**:

Add the securityContext to crd definition and apply it. 

**Anything else we need to know?**:

**Comment by ksatchit** [comment]
@golthitarun , the securitycontext settings are expected to be applied on the resulting experiment pod.

**Comment by golthitarun** [comment]
will the experiment job that gets created have the securityContext added ? or is there any way I can add securityContext as a global config to litmus. 

This is the pod description, not sure if there is anyway to add securitycontext to this job 

```
Name:         pod-delete-o1egk4-j4x2t
Namespace:    litmus
Priority:     0
Node:        xxxxxxx
Start Time:   Fri, 04 Jun 2021 16:54:45 +0000
Labels:       app.kubernetes.io/component=experiment-job
              app.kubernetes.io/part-of=litmus
              app.kubernetes.io/version=1.13.5
              chaosUID=59b89674-9fe2-4116-b6b8-6bad89e3d3ae
              controller-uid=d6d2354e-3008-4f53-9b9c-cac7a07c9351
              job-name=pod-delete-o1egk4
              name=pod-delete
Annotations:  cni.projectcalico.org/podIP: 10.42.0.24/32
              cni.projectcalico.org/podIPs: 10.42.0.24/32
              kubernetes.io/psp: global-restricted-psp
Status:       Pending
IP:           10.42.0.24
IPs:
  IP:           10.42.0.24
Controlled By:  Job/pod-delete-o1egk4
Containers:
  pod-delete-o1egk4:
    Container ID:  
    Image:         litmuschaos/go-runner:1.13.5
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/bash
    Args:
      -c
      ./experiments -name pod-delete
    State:          Waiting
      Reason:       CreateContainerConfigError
    Ready:          False
    Restart Count:  0
    Environment:
      EXPERIMENT_NAME:                   pod-delete
      CHAOS_NAMESPACE:                   litmus
      PODS_AFFECTED_PERC:                
      STATUS_CHECK_TIMEOUT:              180
      APP_KIND:                          deployment
      ANNOTATION_KEY:                    litmuschaos.io/chaos
      CHAOSENGINE:                       pod-delete5mbkm
      TERMINATION_GRACE_PERIOD_SECONDS:  0
      TOTAL_CHAOS_DURATION:              30
      LIB:                               litmus
      AUXILIARY_APPINFO:                 
      TARGET_PODS:                       
      SEQUENCE:                          parallel
      LIB_IMAGE_PULL_POLICY:             Always
      RAMP_TIME:                         
      FORCE:                             false
      CHAOS_INTERVAL:                    10
      APP_NAMESPACE:                     nginx
      CHAOS_UID:                         59b89674-9fe2-4116-b6b8-6bad89e3d3ae
      ANNOTATION_CHECK:                  false
      STATUS_CHECK_DELAY:                2
      JOB_CLEANUP_POLICY:                retain
      APP_LABEL:                         app=nginx
      POD_NAME:                          pod-delete-o1egk4-j4x2t (v1:metadata.name)
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from litmus-admin-token-qx667 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  litmus-admin-token-qx667:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  litmus-admin-token-qx667
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age                 From               Message
  ----     ------     ----                ----               -------
  Normal   Scheduled  108s                default-scheduler  Successfully assigned litmus/pod-delete-o1egk4-j4x2t to avular-opcc-iso-2
  Normal   Pulled     107s                kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 691.397514ms
  Normal   Pulled     106s                kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 474.675609ms
  Normal   Pulled     92s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 503.726487ms
  Normal   Pulled     77s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 557.184056ms
  Normal   Pulled     64s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 538.631876ms
  Normal   Pulled     52s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 508.116789ms
  Normal   Pulled     41s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 554.944511ms
  Warning  Failed     27s (x8 over 107s)  kubelet            Error: container has runAsNonRoot and image has non-numeric user (litmus), cannot verify user is non-root (pod: "pod-delete-o1egk4-j4x2t_litmus(dee4f8bb-344a-4246-b149-884945a072de)", container: pod-delete-o1egk4)
  Normal   Pulled     27s                 kubelet            Successfully pulled image "litmuschaos/go-runner:1.13.5" in 468.865145ms
  Normal   Pulling    13s (x9 over 107s)  kubelet            Pulling image "litmuschaos/go-runner:1.13.5"
```

**Comment by ksatchit** [comment]
In case you'd like a global config change youncan use PSPs: https://docs.litmuschaos.io/docs/litmus-psp/

**Comment by uditgaurav** [comment]
Hi, @golthitarun @ksatchit I think this issue was resolved by using `x-kubernetes-preserve-unknown-fields: true` in the securityContext field of chaosexperiment CRD.

```yaml
 securityContext:
     x-kubernetes-preserve-unknown-fields: true
     type: object
```
Can apply this CRD and try once.
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: chaosexperiments.litmuschaos.io
spec:
  group: litmuschaos.io
  names:
    kind: ChaosExperiment
    listKind: ChaosExperimentList
    plural: chaosexperiments
    singular: chaosexperiment
  scope: Namespaced
  versions:
  - name: v1alpha1
    schema:
      openAPIV3Schema:
        type: object
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
            type: string
          description: 
            type: object
            additionalProperties:
              type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          status:
            x-kubernetes-preserve-unknown-fields: true
            type: object            
          spec:
            type: object
            properties:
              definition:
                x-kubernetes-preserve-unknown-fields: true
                type: object
                properties:
                  args:
                    type: array
                    items:
                      type: string
                  command:
                    type: array
                    items:
                      type: string
                  env:
                    type: array
                    items:
                      type: object
                      properties:
                        name:
                          type: string
                        value:
                          type: string
                  image:
                    type: string
                  imagePullPolicy:
                    type: string
                  labels:
                    type: object
                    additionalProperties:
                      type: string
                  scope:
                    type: string
                    pattern: ^(Namespaced|Cluster)$
                  permissions:
                    type: array
                    items:
                      type: object
                      minProperties: 3
                      required:
                        - apiGroups
                        - resources
                        - verbs
                      properties:
                        apiGroups:
                          type: array
                          items:
                            type: string
                        resources:
                          type: array
                          items:
                            type: string
                        verbs:
                          type: array
                          items:
                            type: string
                        resourceNames:
                          type: array
                          items:
                            type: string
                        nonResourceURLs:
                          type: array
                          items:
                            type: string
                  configMaps:
                    type: array
                    items:
                      type: object
                      minProperties: 2
                      properties:
                        name:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                        mountPath:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                  secrets:
                    type: array
                    items:
                      type: object
                      minProperties: 2
                      properties:
                        name:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                        mountPath:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                  hostFileVolumes:
                    type: array
                    items:
                      type: object
                      minProperties: 3
                      properties:
                        name:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                        mountPath:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                        nodePath:
                          type: string
                          allowEmptyValue: false
                          minLength: 1
                  securityContext:
                    x-kubernetes-preserve-unknown-fields: true
                    type: object
                  hostPID:
                    type: boolean

    served: true
    storage: true
    subresources: {}
  conversion:
    strategy: None
```

**Comment by golthitarun** [comment]
@uditgaurav thanks that works

**Comment by uditgaurav** [comment]
Closing this as the fix is now available with [1.13.6 installation](https://litmuschaos.github.io/litmus/litmus-operator-v1.13.6.yaml).
